crypto/ec/ec_asn1.c

```diff
@@ -876,7 +876,11 @@ EC_GROUP *d2i_ECPKParameters(EC_GROUP **a, const unsigned char **in, long len)
     const unsigned char *p = *in;
 
     if ((params = d2i_ECPKPARAMETERS(NULL, &p, len)) == NULL) {
-        ECerr(EC_F_D2I_ECPKPARAMETERS, EC_R_D2I_ECPKPARAMETERS_FAILURE);
+        int err = ERR_peek_last_error();
+
+        if (ERR_GET_LIB(err) != ERR_LIB_ASN1
+                || ERR_GET_REASON(err) != ERR_R_NESTED_ASN1_ERROR)
+            ECerr(EC_F_D2I_ECPKPARAMETERS, EC_R_D2I_ECPKPARAMETERS_FAILURE);
```

Comment:
This seems dodgy...

Reply:
Various intermediate functions tend to add effectively something like "oh, the function I called returned an error, let me add my own comment on this" with typically little wisdom on the actual error.
This often causes confusion, and in this case it hides the lower-level `ERR_R_NESTED_ASN1_ERROR`
such that in `store_result.c` the check in the following macro goes wrong:
```
#define CLEAR_ERR_MARK()                                                \
    do {                                                                \
        int err = ERR_peek_last_error();                                \
                                                                        \
        if (ERR_GET_LIB(err) == ERR_LIB_ASN1                            \
            && ERR_GET_REASON(err) == ERR_R_NESTED_ASN1_ERROR)          \
            ERR_pop_to_mark();                                          \
        else                                                            \
            ERR_clear_last_mark();                                      \
    } while(0)

```
It took me quite some effort to find this out and come up with a good solution.

Reply:
> Various intermediate functions tend to add effectively something like "oh, the function I called returned an error, let me add my own comment on this" with typically little wisdom on the actual error.

I agree, the diverse "ERR_R_XXX_LIB" are exactly that sort of thing, and I personally think they should be abolished.  And granted, this one is of the same type, as it just says "this function I called failed for some reason"...  which most probably indicates that there's a chance that the function it called might not add an error record of its own, which I would say is a bug in most cases (some very internal even though public utility functions might get away with that, but something as central as ASN.1 should at all times add error records when there's an error).

What feels dodgy here is that it introduces some randomness.  This error might or might not be added, and anyone thinking this will be trustable, even within the same OpenSSL version, will be put to shame.

What I would do if I was you, is simply to remove this error record, unconditionally.  It doesn't really fulfil any purpose other than a stack trace point, and a real debugger is better suited for that...

Reply:
I fully agree with your comments, in particular to dislike purely verbose and defensive error queue additions, which only cause trouble.

Following your suggestion I've removed unconditionally all those additions that I found to be harmful for the diagnostics of `decoder_process()`.